internal/api/client.go

```diff
@@ -52,3 +61,120 @@ func NewClient(cfg ClientConfig) *Client {
 		httpClient:       httpClient,
 	}
 }
+
+var (
+	retryTimeout = 130 * time.Second
+	initialDelay = 3000 * time.Millisecond
+)
+
+var ErrRetryTimeout = errors.New("request retry timeout")
+
+type errorResponse struct {
+	Message string `json:"message"`
+}
+
+type httpRequest struct {
+	Method string
+	URL    string
+	Body   any
+}
+
+// DoWithRetry sends http request with retries.
+// Successful API response (status code 200) is JSON decoded and stored in the value pointed to by v.
+// The request will be retried when the server returns 429 or 5xx status code, or when there is a network error.
+// After reaching the retry timeout, the function will return ErrRetryTimeout.
+// The request will not be retried when the server returns 4xx status code,
+// and the error message will be returned as an error.
+func (c *Client) DoWithRetry(ctx context.Context, req httpRequest, v interface{}) (*http.Response, error) {
+	r := roko.NewRetrier(
+		roko.TryForever(),
+		roko.WithStrategy(roko.ExponentialSubsecond(initialDelay)),
+		roko.WithJitter(),
+	)
+
+	retryContext, cancelRetryContext := context.WithTimeout(ctx, retryTimeout)
+	defer cancelRetryContext()
+
+	// retry loop
+	debug.Printf("Sending request %s %s", req.Method, req.URL)
+	resp, err := roko.DoFunc(retryContext, r, func(r *roko.Retrier) (*http.Response, error) {
+		if r.AttemptCount() > 0 {
+			debug.Printf("Retrying requests, attempt %d", r.AttemptCount())
+		}
+
+		reqBody, err := json.Marshal(req.Body)
+		if err != nil {
+			r.Break()
+			return nil, fmt.Errorf("converting body to json: %w", err)
+		}
+
+		// Each request times out after 15 seconds, chosen to provide some
+		// headroom on top of the goal p99 time to fetch of 10s.
+		reqContext, cancelReqContext := context.WithTimeout(ctx, 15*time.Second)
+		defer cancelReqContext()
+
+		req, err := http.NewRequestWithContext(reqContext, req.Method, req.URL, bytes.NewBuffer(reqBody))
+		if err != nil {
+			r.Break()
+			return nil, fmt.Errorf("creating request: %w", err)
+		}
+		req.Header.Add("Content-Type", "application/json")
+
+		resp, err := c.httpClient.Do(req)
+
+		// If we get an error before getting a response,
+		// which means there is a network error (e.g. protocol error, timeout),
+		// we should return and retry.
+		if err != nil {
+			debug.Printf("Error sending request: %v", err)
+			return nil, err
+		}
+
+		debug.Printf("Response code %d", resp.StatusCode)
+
+		// If we get a 429, we should return and retry after the rate limit resets.
+		if resp.StatusCode == http.StatusTooManyRequests {
+			if rateLimitReset, err := strconv.Atoi(resp.Header.Get("RateLimit-Reset")); err == nil {
```

Comment:
@wooly can you confirm that this is what the API will return in the header when API rate limit is reached.

Reply:
Yep, that's it.